### PR TITLE
runtime: implement a receipt preparation pipeline

### PR DESCRIPTION
**Pull request by nagisa**
This PR implements a very basic form of pipelining of the contract preparation. In particular with this PR only actions from one upcoming "interesting" local receipt is handled, with other receipt kinds (incoming, delayed) being left for a future change. I don't anticipate any trouble integrating them as they largely have similar data and structure as the local receipts seen here.

The primary reason I want this reviewed separately, however, is due to the changes to storage. You will notice that I have largely changed the code to read contracts without going through The State (i.e. it now reads the contract code directly via the code hash.) However at the same time writing of the contract continues involving The State; largely because I don't want to deal with migrations and similar such complexities… but also because `TrieStorage` does not provide write access. This also means that writing is transactional and I had to add some work-arounds to the `store::contract::Storage` in order to make deployed contracts visible before the commit.

### CODE REVIEW

File: runtime/runtime/src/pipelining.rs
Comment (nagisa) [inline runtime/runtime/src/pipelining.rs:58]: Question to the storage folks: is this the right way to get access to the contracts here? If not, what would be the right way? I feel like this code needn't be concerned about recording and such, because we already handle it separately through prefetching. The access mechanism has to be thread-safe.

To give some more background, the implementation here is meant to access the storage for the previous block than the one currently processed, largely because sharing `TrieUpdate` across threads is not really easy to think about. So the idea is to directly access RocksDB (or another implementation of storage), which would have changes from the previous block/chunk already committed.

File: runtime/runtime/src/lib.rs
Comment (nagisa) [inline runtime/runtime/src/lib.rs:2300]: Question to storage folks: what's the right way to access the underlying storage here? If there isn't one, what do you think would be the right way to expose it to this code?

**Comment by nagisa** [comment]
r? @Ekleog-NEAR for the runtimeish bits
r? @pugachAG for the storagey bits

This one I think is better reviewed as a whole PR (files are nicely separated and commits -- less so.) Feel free to note if you think any of the FIXMEs or TODOs are critical to resolve before this can land.

Side note: I'm surprised by how good our existent coverage is. Most of the uncovered lines are actually just panics :D

File: runtime/runtime/src/pipelining.rs
Comment (Ekleog-NEAR) [inline runtime/runtime/src/pipelining.rs:261]: Nit: set to finished before actually calling `prepare_function_call`, and release the status guard, so that the mutex is held for less time? It might help if the prefetcher thread is coming to prefetch this contract just when the compute thread is handling it.

**Review by Ekleog-NEAR — CHANGES_REQUESTED** [review]
I reviewed all the files except those in `core/store`.

Only one important comment that deserved the changes requested: the issue if one receipt were to contain two deploycontract actions.

Aside from that, the code looks nice! :)

File: runtime/runtime/src/pipelining.rs
Comment (nagisa) [reply runtime/runtime/src/pipelining.rs:154]: I don't think we want to have a separate thread pool, no. We already have an excess of thread pools with NCPU threads in each. But even if the pool is occupied with other business, the implementation here won't wait on it unnecessarily.

File: runtime/runtime/src/pipelining.rs
Comment (nagisa) [reply runtime/runtime/src/pipelining.rs:131]: If I understand the concern correctly -- it is fine currently. 

1. All deploys within a single receipt (which is what this is processing) affect only the receipt receiver account which is the same across all actions within that receipt;
2. Once an account is blocked, it remains blocked for the remainder of a chunk, so it doesn't matter if there are further deployments happening to the same account once it is blocked.

File: runtime/runtime/src/actions.rs
Comment (nagisa) [reply runtime/runtime/src/actions.rs:629]: Adjusted the wording.

File: runtime/runtime/src/pipelining.rs
Comment (pugachAG) [inline runtime/runtime/src/pipelining.rs:139]: I guess we don't want warning here since having multiple function calls in the same receipt is normal and it would trigger that case, right?

File: runtime/runtime/src/lib.rs
Comment (pugachAG) [inline runtime/runtime/src/lib.rs:1667]: what if the first receipt contains deploy contract action, should we block the account_id for further pipelining in that case?

**Review by pugachAG — COMMENTED** [review]
storage parts LGTM, got a small question regarding the first receipt

File: runtime/runtime/src/pipelining.rs
Comment (Ekleog-NEAR) [reply runtime/runtime/src/pipelining.rs:154]: My thinking was actually the opposite: it seems to me that this pipelining prefetch business could be busying the thread pool for other, more pressing business. And by limiting the pipelining prefetch thread pool to 1 thread, we could also make sure it doesn't take up too much compute.

This being said, feel free to ignore the suggestion, it's most likely not going to be anything critical :)

File: runtime/runtime/src/pipelining.rs
Comment (Ekleog-NEAR) [reply runtime/runtime/src/pipelining.rs:131]: Oh yes I had forgotten point 1, thank you for the reminder :)

**Review by Ekleog-NEAR — APPROVED** [review]
r+ for the runtime bits :)

File: runtime/runtime/src/pipelining.rs
Comment (nagisa) [reply runtime/runtime/src/pipelining.rs:139]: The key used in the map is receipt hash + action index, so this case being hit would mean that the caller is submitting the same receipt (with all its actions) multiple times in the same chunk, possibly with the intent to execute the actions multiple times. This won't work in the current implementation as the prepared contracts cannot be reused.

File: runtime/runtime/src/pipelining.rs
Comment (nagisa) [reply runtime/runtime/src/pipelining.rs:154]: Fair enough, I realize now that this is actually more aggressive in scheduling new work than I meant it to be, so I'll need to rewrite this anyhow. Limiting the pool, however, isn't quite the right approach IMO.

File: runtime/runtime/src/lib.rs
Comment (nagisa) [reply runtime/runtime/src/lib.rs]: Good eye! I think I addressed this by not trying to be as smart. I should add a regression test for this too (please do not resolve this thread quite yet.)

@Ekleog-NEAR I also made changes which I think will make it clearer when and how many receipts are submitted in parallel, so the currently last commit should resolve your concerns with that too?

**Review by pugachAG — APPROVED** [review]

File: runtime/runtime/src/lib.rs
Comment (Ekleog-NEAR) [reply runtime/runtime/src/lib.rs]: The changes LGTM :)

**Comment by akhi3030** [comment]
CC: @https://github.com/near/nearcore/issues/11319

**Comment by nagisa** [comment]
@race-of-sloths include

**Comment by race-of-sloths** [comment]
@nagisa Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
New Sloth joined the Race! Welcome!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/nagisa?type=bot&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/nagisa?type=bot&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/nagisa)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command. Alternatively, autoscoring [1,2] will be applied for this pull request

</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/nagisa)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

**Comment by akhi3030** [comment]
@race-of-sloths score 13

**Pull request merged**
